### PR TITLE
Add support for models compiled through dynamo / inductor

### PR DESCRIPTION
**Pull request by lantiga**
## What does this PR do?

This PR introduces dynamo support to Lightning. A prior design proposed the use of a callback.

However, since compilation is going to be first-class in PyTorch, here we add a `compile` method to `LightningModule` so that users can do

```python
model = BoringModel()
model.compile(...)
```

and control compilation this way by passing arguments as exposed by `torch.compile` (https://pytorch.org/docs/master/generated/torch.compile.html).

Alternatively, the user can activate compilation by passing a `compile=True` flag to the trainer (or a dict with options for granular control).

```python
model = BoringModel()
trainer = Trainer(compile=True)
trainer.fit(model)
```

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: src/pytorch_lightning/core/module.py [hunk 1]
```diff
@@ -270,6 +271,11 @@ def loggers(self) -> List[Logger]:
         """Reference to the list of loggers in the Trainer."""
         return self.trainer.loggers if self._trainer else []
 
+    @property
+    def compiler(self) -> Optional[str]:
```
Comment (tchaton) [inline src/pytorch_lightning/core/module.py:275]: I would have imagined a slightly different approach.

Either introduce a new hook:

```python
from lightning.pytorch import Compiler

def configure_compiler(self):
     return Compiler(backend="inductor", nopython=True, dynamic=True)
```

By doing so, if the users override this new hook, we would directly compile the model for them at the right level.

If a user is using DDP, I would expect the entire training_step to be wrapped including the forward method of the DDP wrapper.

or 

```python
Trainer(

   compile=Compiler(backend="inductor", nopython=True, dynamic=True)
)
```

Comment (lantiga) [reply src/pytorch_lightning/core/module.py:275]: As discussed, I'm going to change the design to something different altogether so that we can just have users rely on `torch.compile`

File: src/pytorch_lightning/core/module.py [hunk 2]
```diff
@@ -1786,6 +1792,66 @@ def _verify_is_manual_optimization(self, fn_name: str) -> None:
...
+        """
+        try:
+            import torch._dynamo as dynamo
+            from torch._dynamo.eval_frame import lookup_backend
+            from torch._inductor.config import InductorConfigContext
+        except ImportError as e:
+            raise ImportError(
+                "Please make sure you are running on PyTorch version 1.14+, "
+                "or a PyTorch nightly published after Fri 3 Dec 2022."
+            ) from e
```
Comment (Borda) [inline src/pytorch_lightning/core/module.py:1838]:
```suggestion
        except ImportError as err:
            raise ImportError(
                "Please make sure you are running on PyTorch version 1.14+, "
                "or a PyTorch nightly published after Fri 3 Dec 2022."
            ) from err
```

File: src/pytorch_lightning/core/module.py [hunk 3]
```diff
@@ -1786,6 +1792,66 @@ def _verify_is_manual_optimization(self, fn_name: str) -> None:
...
+
+            def _compile_fn(fn: Callable, inputs_: Any) -> Any:
+                with cm:
+                    return compile_fn(fn, inputs_)
+
+            _compile_fn._torchdynamo_orig_callable = compile_fn  # type: ignore[attr-defined]
+            backend = _compile_fn
+
+        optimize = dynamo.optimize(backend=backend, nopython=fullgraph, dynamic=dynamic, **kwargs)
+        self.forward = optimize(self.forward)  # type: ignore[assignment]
```
Comment (Borda) [inline src/pytorch_lightning/core/module.py:1852]:
```suggestion
        optimizer = dynamo.optimize(backend=backend, nopython=fullgraph, dynamic=dynamic, **kwargs)
        self.forward = optimizer(self.forward)  # type: ignore[assignment]
```

File: src/pytorch_lightning/trainer/trainer.py [hunk 4]
```diff
@@ -389,6 +390,11 @@ def __init__(
 
             inference_mode: Whether to use :func:`torch.inference_mode` or :func:`torch.no_grad` during
                 evaluation (``validate``/``test``/``predict``).
+
+            compile: Whether to optimize the model using a compiler.
+                It accepts a boolean, or a dict for granular control over optimizations containing the
+                arguments to :meth:`~pytorch_lightning.core.module.LightningModule.compile` in key-value form.
+                Default: ``False``.
```
Comment (Borda) [inline src/pytorch_lightning/trainer/trainer.py:397]:
```suggestion
```
already in func footprint

File: src/pytorch_lightning/core/module.py [hunk 5]
```diff
@@ -1786,6 +1792,66 @@ def _verify_is_manual_optimization(self, fn_name: str) -> None:
...
+    ) -> None:
+        """Optimizes the model using the Dynamo compiler.
+
+        Args:
+            backend (str or Callable): compiler backend to be used. Default: ``"inductor"``.
+                Call ``torch._dynamo.list_backends()`` to get a full list.
+            dynamic (bool): Use dynamic shape tracing. Default: ``False``.
+            fullgraph (bool): Whether it is ok to break model into several subgraphs. Default: ``False``.
+            mode (str): Can be either ``"default"``, ``"reduce-overhead"`` or ``"max-autotune"``.
+                Only relevant for the "inductor" backend.
```
Comment (Borda) [inline src/pytorch_lightning/core/module.py:1811]:
```suggestion
            backend: compiler backend to be used.
                Call ``torch._dynamo.list_backends()`` to get a full list.
            dynamic: Use dynamic shape tracing.
            fullgraph: Whether it is ok to break model into several subgraphs.
            mode: Can be either ``"default"``, ``"reduce-overhead"`` or ``"max-autotune"``.
                Only relevant for the "inductor" backend.
```
already in the method typing, so in docs too

File: src/pytorch_lightning/trainer/trainer.py [hunk 6]
```diff
@@ -493,6 +499,14 @@ def __init__(
 
         self._inference_mode: bool = inference_mode
 
+        if not (type(compile) == bool or isinstance(compile, dict)):
```
Comment (Borda) [inline src/pytorch_lightning/trainer/trainer.py:502]:
```suggestion
        if not isinstance(compile, (bool, dict)):
```

**Review by Borda — APPROVED** [review]

**Comment by lantiga** [comment]
I forced pushed too much, GitHub is getting confused. Opening a separate PR.
